tests/h/search/query_test.py

```diff
@@ -622,14 +637,54 @@ def search(self, search):
         return search
 
 
-@pytest.mark.usefixtures('pyramid_config')
-class TestNipsaFilter(object):
+@pytest.mark.usefixtures('pyramid_config', 'moderation_service')
+class TestHiddenFilter(object):
+
+    @pytest.mark.parametrize('nipsa_val,hidden_val,show_val', [
+        # both nipsa and hidden fields are set, so hide the annotation
+        (True, True, False),
+        # only nipsa field is set, still hide the annotation
+        (True, False, False),
+        # only hidden field is set, still hide the annotation
+        (False, True, False),
+        # neither field is set, don't hide the annotation
+        (False, False, True),
+    ])
+    def test_visibility_of_moderated_and_nipsaed_annotations(
+        self, index, factories, pyramid_request, search, user,
+        AnnotationSearchIndexPresenter, nipsa_val, hidden_val, show_val
+    ):
+
+        pyramid_request.user = user
+        search.append_modifier(query.HiddenFilter(pyramid_request))
+
+        annotation_1 = factories.Annotation.build()
+        annotation_2 = factories.Annotation.build(userid=user.userid)
+        expected_ids = [annotation_2.id]
+
+        if show_val:
+            expected_ids.append(annotation_1.id)
+
+        presenter = AnnotationSearchIndexPresenter.return_value
```

Comment:
I think the approach here works well on a practical level because otherwise it would be necessary to jump through a bunch of hoops to mark the annotation as NIPSA'd / hidden.

Generally our tests fall into two buckets, those which mock everything non-trivial (most unit tests) and those which mock nothing (functional tests for the whole application). The tests in this module are actually neither unit nor functional but rather integrated tests for a subset of functionality in the application. This is _useful_ but we don't have clearly established conventions for them.

Reply:
Related: @lyzadanger shared a good talk about mocking recently which is connected to this: https://hypothes-is.slack.com/archives/C4K6M7P5E/p1534901732000100

Reply:
Yep - I agree with what both Hannah and Rob are saying here.

So this is a test for `HiddenFilter` and it mocks `AnnotationSearchIndexPresenter` and some services? That's the right approach -- the logic inside presenters and services is tested by those presenters and services' own unit tests, and the unit tests for `HiddenFilter` should test only `HiddenFilter`'s own logic and that it uses the presenters and services correctly.

If you're interested in more details about the reasoning behind this approach and when and when not to mock, we have some notes on our approach to mocking in our docs here: <https://h.readthedocs.io/en/latest/developing/testing/>. Those notes also link to the series of Python testing posts on my blog, which includes one [post about when and when not to mock](https://www.seanh.cc/posts/when-to-use-mocks).

The idea of integrated tests that are not entirely functional or end-to-end tests, but that integrate a subset of modules within our code and test them together, is interesting. As Rob says, we don't have a place for such tests in our codebase right now. `tests/h` contains unit tests, and `tests/functional` contains end-to-end functional tests. I would definitely be **against writing integrated tests in `tests/h`** because it breaks the consistent pattern of what our approach to testing is and what is and isn't mocked. But I'm not against the idea of having integrated tests if they could be put in `tests/integrated` or somewhere. We can get pretty far with just the two extremes of unit and functional tests, though.

Reply:
👍